### PR TITLE
osd: Corrupt objects stop snaptrim and mark pg snaptrim_error

### PR DESCRIPTION
**Pull request by dzafman**

### CODE REVIEW

**Comment by dzafman** [comment]
In Progress

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -13963,16 +13978,21 @@ boost::statechart::result PrimaryLogPG::AwaitAsyncWork::react(const DoSnapWork&)
...
       ldout(pg->cct, 10) << "could not get write lock on obj "
 			 << object << dendl;
       if (in_flight.empty()) {
+        assert(error == -ENOLCK);
 	ldout(pg->cct, 10) << "waiting for it to clear"
 			   << dendl;
 	return transit< WaitRWLock >();
 
       } else {
+	// XXX: This can't happen anymore
```
Comment (liewegas) [inline src/osd/PrimaryLogPG.cc]: remove it then?

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -9620,7 +9626,11 @@ ObjectContextRef PrimaryLogPG::get_object_context(
     dout(10) << __func__ << ": creating obc from disk: " << obc
 	     << dendl;
   }
-  assert(obc->ssc);
+
+  // XXX: Caller doesn't expect this
+  if (obc->ssc == NULL)
+    return ObjectContextRef();   // -ENOENT!
```
Comment (liewegas) [inline src/osd/PrimaryLogPG.cc]: maybe derr here?  it makes me a bit nervous to turn this into silently an ENOENT when we can't load the snapset.

**Comment by dzafman** [comment]
Remaining issue: If a snaptrim is missing and repair fixes it, the object context cache isn't fixed, so that snap trimmer still can't complete.  A reboot of the osd cleans up the pg.

**Comment by dzafman** [comment]
retest this please

**Comment by dzafman** [comment]
dzafman-2017-06-26_14:07:20-rados-wip-13837-distro-basic-smithi

Testing passed.

The Shaman "Ubuntu Xenial notcmalloc" build failed, thus the failed package fetches

Failed
1328151 Failed to fetch package
1328164	"cephtest/swift/test/functional   -v -a ''!fails_on_rgw''" seen in master
1328203 Out of space Tracker #20422
1328248	Failed to fetch package
1328254	cp: cannot stat ‘/var/log/audit/audit.log’: No such file or directory
1328282	Out of space Tracker #20422
1328316	Out of space Tracker #20422
1328339	Failed to fetch package
1328360 Failed to fetch package

Dead
1328310 Out of space tracker #20422
1328370 Stuck waiting for clean early on. Tracker #20439

**Pull request merged**